### PR TITLE
LWDEV-7668 Redis persistance: Processed messages ids

### PR DESCRIPTION
**Pull request by rocketmannot**

### CODE REVIEW

File: src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt [hunk 1]
```diff
@@ -31,7 +40,6 @@ class RedisPersistenceManager(
             persistData(data)
             true
         } catch (e: Exception) {
-            jedisHolder.fail()
```
Comment (rocketmannot) [inline src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt]: Please note. i removed this flag as it looks like we dont need it

**Comment by rocketmannot** [comment]
NOTE: changes in settings json required

File: src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt (hunk 1 above)
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt]: decided to revert changes and go with extenralfail flag - as roll back operation is risky and should be hit only in exception cases

**Comment by barsukovskii** [comment]
When i tested locally i saw next exception during application shutdown:

Exception in thread "com.lykke.matching.engine.database.redis.monitoring.impl.RedisHealthStatusHolderImpl" java.lang.IllegalStateException: org.springframework.context.annotation.AnnotationConfigApplicationContext@f926e6 has been closed already
	at org.springframework.context.support.AbstractApplicationContext.assertBeanFactoryActive(AbstractApplicationContext.java:1069)
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1084)
	at org.springframework.context.event.ApplicationListenerMethodAdapter.getTargetBean(ApplicationListenerMethodAdapter.java:291)
	at org.springframework.context.event.ApplicationListenerMethodAdapter.doInvoke(ApplicationListenerMethodAdapter.java:261)
	at org.springframework.context.event.ApplicationListenerMethodAdapter.processEvent(ApplicationListenerMethodAdapter.java:182)
	at org.springframework.context.event.ApplicationListenerMethodAdapter.onApplicationEvent(ApplicationListenerMethodAdapter.java:144)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:399)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:366)
	at com.lykke.matching.engine.database.redis.monitoring.impl.RedisHealthStatusHolderImpl$init$$inlined$fixedRateTimer$1.run(Timer.kt:147)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)

File: src/main/kotlin/com/lykke/matching/engine/config/spring/DatabaseAccessorConfig.kt [hunk 2]
```diff
@@ -46,34 +35,72 @@ open class DatabaseAccessorConfig {
...
-    open fun jedisHolder(): JedisHolder {
-        return when (config.me.walletsStorage) {
-            WalletsStorage.Azure -> EmptyJedisHolder()
-            WalletsStorage.Redis -> {
-                val holder = balancesDatabaseAccessorsHolder()
-                DefaultJedisHolder(holder.redisConfig)
-            }
+    open fun readOnlyProcessedMessagesDatabaseAccessor(jedisPool: JedisPool): ReadOnlyProcessedMessagesDatabaseAccessor {
+        return when (config.me.storage) {
+            Storage.Azure -> fileProcessedMessagesDatabaseAccessor()
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/config/spring/DatabaseAccessorConfig.kt]: yeah - we had conversation with Dmitry abou that. This was the initial idea - however in order to implement that we need some kind of separate persistence manager for processed message - we decided that this extra effort will take some time - so decided to go with one flag for balances and messages.

File: src/main/kotlin/com/lykke/matching/engine/database/redis/monitoring/RedisHealthStatusHolder.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.lykke.matching.engine.database.redis.monitoring
+
+import com.lykke.matching.engine.utils.monitoring.HealthMonitor
+
+interface RedisHealthStatusHolder: HealthMonitor {
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/monitoring/RedisHealthStatusHolder.kt]: removed

File: src/main/kotlin/com/lykke/matching/engine/database/redis/monitoring/impl/RedisHealthStatusHolderImpl.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    @PostConstruct
+    open fun init() {
+        if (config.me.storage != Storage.Redis) {
+            return
+        }
+        fixedRateTimer(RedisHealthStatusHolderImpl::class.java.name, false, 0, updateInteval) {
+            val redisAlive = isRedisAlive()
+            ok = externalFail || redisAlive
+            applicationEventPublisher.publishEvent(HealthMonitorEvent(ok, MonitoredComponent.REDIS))
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/monitoring/impl/RedisHealthStatusHolderImpl.kt]: suggest to leave as is - publish event is cheap operation. in other case i should handle ok previous state

File: src/main/kotlin/com/lykke/matching/engine/order/GenericLimitOrderProcessor.kt [hunk 5]
```diff
@@ -43,30 +44,33 @@ class GenericLimitOrderProcessor(private val limitOrderService: GenericLimitOrde
             applicationSettingsCache,
             LOGGER)
 
-    fun checkAndProcessStopOrder(messageId: String, assetPairId: String, now: Date) {
+    fun checkAndProcessStopOrder(messageId: String, processedMessage: ProcessedMessage?, assetPairId: String, now: Date) {
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/order/GenericLimitOrderProcessor.kt]: suggest to leave as is - very complicated call stack - i did not managed to trace if processedMessage is saved in all code paths - as it is too complicated. For now i just pass processed message in every place where we perform RedisPersistenceManager#persist operation - if in some cases we duplicate processedMessage persistence - no problem as we are using set at the redis level. this part will be rewritten after we will move to separate persist stage

File: src/main/kotlin/com/lykke/matching/engine/database/redis/monitoring/impl/RedisHealthStatusHolderImpl.kt [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    private fun isRedisAlive(): Boolean {
+        try {
+            jedisPool.resource.use { jedis ->
+                val transaction = jedis.multi()
+                try {
+                    transaction.select(getPingDb())
+                    transaction[PING_KEY] = PING_VALUE
+                    transaction.exec()
+                } catch (e: Exception) {
+                    LOGGER.error("Redis ping operation failed")
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/monitoring/impl/RedisHealthStatusHolderImpl.kt]: done

File: src/main/kotlin/com/lykke/matching/engine/database/redis/monitoring/impl/RedisHealthStatusHolderImpl.kt [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    var ok = false
+
+    override fun ok() = ok && !externalFail
+
+    override fun fail() {
+        externalFail = true
+    }
+
+    private fun getPingDb(): Int {
+        return if (config.me.redis.balanceDatabase != 0) 0 else 1
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/monitoring/impl/RedisHealthStatusHolderImpl.kt]: added new parameter for PingDB at settings json

File: src/main/kotlin/com/lykke/matching/engine/database/redis/accessor/impl/RedisProcessedMessagesDatabaseAccessor.kt [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                    .map {conf.asObject(it.toByteArray()) as ProcessedMessage}
+                    .collect(Collectors.toSet())
+        }
+     }
+
+    fun save(transaction: Transaction, message: ProcessedMessage) {
+        transaction.select(dbIndex)
+        val key = getKey(message)
+        SetUtils.performAtomicSaveSetExpire(transaction, key, conf.asJsonString(message), timeToLive)
+        LOGGER.debug("Persist of processed message: $message to redis performed")
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/accessor/impl/RedisProcessedMessagesDatabaseAccessor.kt]: set Trace level

File: src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt [hunk 9]
```diff
@@ -66,7 +78,28 @@ class RedisPersistenceManager(
...
+
+        redisProcessedMessagesDatabaseAccessor.save(transaction, processedMessage)
+    }
+
+    private fun persistBalances(transaction: Transaction, assetBalances: Collection<AssetBalance>?) {
+        if (assetBalances == null) {
+            return
+        }
+
+        LOGGER.info("Start to persist balances in redis")
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt]: done

File: src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt [hunk 10]
```diff
@@ -66,7 +78,28 @@ class RedisPersistenceManager(
         }
     }
 
-    private fun init() {
+    private fun persistProcessedMessages(transaction: Transaction, processedMessage: ProcessedMessage?) {
+        LOGGER.info("Start to persist processed messages in redis")
+
+        if (processedMessage == null) {
+            LOGGER.debug("Processed message is empty, skip persisting")
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt]: i have set trace level - looks like we have debug level on for even production env .
For dev logs we should use trace -level as debug is turned on at the prod env.

File: src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt [hunk 11]
```diff
@@ -66,7 +78,28 @@ class RedisPersistenceManager(
         }
     }
 
-    private fun init() {
+    private fun persistProcessedMessages(transaction: Transaction, processedMessage: ProcessedMessage?) {
+        LOGGER.info("Start to persist processed messages in redis")
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/redis/RedisPersistenceManager.kt]: set Trace level

**Pull request merged**